stan/math/rev/functor/operands_and_partials.hpp

```diff
@@ -115,59 +145,43 @@ class operands_and_partials<Op1, Op2, Op3, Op4, Op5, var> {
-        edge5_.container_operands());
-    auto container_partials = std::tuple_cat(
-        edge1_.container_partials(), edge2_.container_partials(),
-        edge3_.container_partials(), edge4_.container_partials(),
-        edge5_.container_partials());
-
-    return var(return_vari(value, edges_size, varis, partials,
-                           container_operands, container_partials));
-  }
-
- private:
-  /**
-   * Deduces types and constructs the vari to return from `build()`.
-   * @param value the value
-   * @param edges_size number of elements in all non-var_value of container
-   * edges
-   * @param varis pointer to varis from non-var_value of container edges
-   * @param partials pointer to values for partials from non-var_value of
-   * container edges
-   * @param container_operands operands from var_value of container edges
-   * @param container_partials partial derivatives from var_value of container
-   * edges
-   */
-  template <typename... Ops, typename... Partials>
-  auto* return_vari(double value, size_t edges_size, vari** varis,
-                    double* partials,
-                    const std::tuple<Ops...>& container_operands,
-                    const std::tuple<Partials...>& container_partials) {
-    return new precomputed_gradients_vari_template<
-        std::tuple<arena_t<Ops>...>, std::tuple<arena_t<Partials>...>>(
-        value, edges_size, varis, partials, container_operands,
-        container_partials);
+    var ret(value);
+    if (!is_constant<Op1>::value) {
+      reverse_pass_callback([operand = to_arena(edge1_.operand()),
+                             partial = to_arena(edge1_.partial()),
+                             ret]() mutable {
+        internal::update_adjoints(operand, partial, ret);
+      });
+    }
+    if (!is_constant<Op2>::value) {
+      reverse_pass_callback([operand = to_arena(edge2_.operand()),
+                             partial = to_arena(edge2_.partial()),
+                             ret]() mutable {
+        internal::update_adjoints(operand, partial, ret);
+      });
+    }
+    if (!is_constant<Op3>::value) {
+      reverse_pass_callback([operand = to_arena(edge3_.operand()),
+                             partial = to_arena(edge3_.partial()),
+                             ret]() mutable {
+        internal::update_adjoints(operand, partial, ret);
+      });
+    }
+    if (!is_constant<Op4>::value) {
+      reverse_pass_callback([operand = to_arena(edge4_.operand()),
+                             partial = to_arena(edge4_.partial()),
+                             ret]() mutable {
+        internal::update_adjoints(operand, partial, ret);
+      });
+    }
+    if (!is_constant<Op5>::value) {
+      reverse_pass_callback([operand = to_arena(edge5_.operand()),
+                             partial = to_arena(edge5_.partial()),
+                             ret]() mutable {
+        internal::update_adjoints(operand, partial, ret);
+      });
```

Comment:
Regardless of what we do here we need to benchmark it to make sure we are not making it slower (including cases like all scalars, small vectors, large vectors).

Reply:
Yes so turns out I was adding about 10-15ns of overhead here until I fixed a few things up

1. Using `make_callback_var()`
2. Adding noexcept to some constructors and member functions (I guess this lets the compiler optimize more aggressively whenever it can't inline something
3. Passing `var` and other trivial types by value!

For the benchmark [here](https://gist.github.com/SteveBronder/079f33629cdfc43875a8a3ad304518e3) I'm testing scalars with one, two, and three parameters of mixed var and double values.

You can run it with perf-math using 

```bash
# in a file called docomp.sh
# and put the benchmark in a folder called callback
cd math
git reset --hard
git pull
git checkout ${1}
git pull
cd ..
rm ./callback/${2}
make ./callback/${2}
./callback/${2}


cd math
git reset --hard
git checkout develop
cd ..
rm ./callback/${2}
make ./callback/${2}
./callback/${2}
```

Then calling 

```
./docomp.sh feature/constexpr-double-ops-partials operands_and_partials
```

I'm getting the following now

```
----------------------------------------------------------------------------------------------
PR Benchmark                                                    Time             CPU   Iterations
----------------------------------------------------------------------------------------------
toss_me<extra_alloc>                                     0.000 ns        0.000 ns            0
operands_and_partial_single/manual_time                   23.0 ns         50.9 ns     30501562
operands_and_partial_double<double, var>                  51.0 ns         51.0 ns     13716475
operands_and_partial_double<var, double>                  50.6 ns         50.6 ns     13706392
operands_and_partial_double<var, var>                     52.1 ns         52.1 ns     13434745
operands_and_partial_triple<double, double, double>       49.3 ns         49.3 ns     14190612
operands_and_partial_triple<var, double, double>          50.7 ns         50.7 ns     13808773
operands_and_partial_triple<double, var, double>          50.9 ns         50.9 ns     13747977
operands_and_partial_triple<double, double, var>          51.0 ns         51.0 ns     13704645
operands_and_partial_triple<double, var, var>             52.2 ns         52.2 ns     13386284
operands_and_partial_triple<var, double, var>             52.8 ns         52.8 ns     13164809
operands_and_partial_triple<var, var, double>             53.3 ns         53.3 ns     13179378
operands_and_partial_triple<var, var, var>                54.7 ns         54.7 ns     12775673
```

```
----------------------------------------------------------------------------------------------
Develop Benchmark                                                    Time             CPU   Iterations
----------------------------------------------------------------------------------------------
toss_me<extra_alloc>                                     0.000 ns        0.000 ns            0
operands_and_partial_single/manual_time                   23.9 ns         51.8 ns     29379580
operands_and_partial_double<double, var>                  52.1 ns         52.1 ns     13507725
operands_and_partial_double<var, double>                  53.0 ns         53.0 ns     13150442
operands_and_partial_double<var, var>                     53.4 ns         53.4 ns     12998046
operands_and_partial_triple<double, double, double>       49.1 ns         49.1 ns     14267525
operands_and_partial_triple<var, double, double>          52.0 ns         52.0 ns     13467392
operands_and_partial_triple<double, var, double>          52.0 ns         52.0 ns     13390830
operands_and_partial_triple<double, double, var>          51.8 ns         51.8 ns     13487857
operands_and_partial_triple<double, var, var>             53.2 ns         53.2 ns     12737199
operands_and_partial_triple<var, double, var>             54.0 ns         54.0 ns     12822179
operands_and_partial_triple<var, var, double>             53.8 ns         53.8 ns     13039172
operands_and_partial_triple<var, var, var>                55.9 ns         55.9 ns     12496295
```

So they are certainly within a hair of one another +/- a few nanoseconds which I think is fine. Do you think I need benchmarks for vectors etc? I would guess those would just be plain faster since we are not requesting all that memory anymore

Reply:
Actually a vector benchmark was pretty trivial to write [bench](https://gist.github.com/SteveBronder/24db8f5b8379c1e4ce4142dc3cbfbb70)

```
-----------------------------------------------------------------------------------------------------------------------
PR Benchmark                                                                          Time             CPU   Iterations
-----------------------------------------------------------------------------------------------------------------------
operands_and_partial_triple_vec<var, double, double>/1024/manual_time_mean         1133 ns         5248 ns           20
operands_and_partial_triple_vec<var, double, double>/1024/manual_time_median       1127 ns         5247 ns           20
operands_and_partial_triple_vec<var, double, double>/1024/manual_time_stddev       20.8 ns         9.40 ns           20
operands_and_partial_triple_vec<var, double, double>/2048/manual_time_mean         2086 ns        10100 ns           20
operands_and_partial_triple_vec<var, double, double>/2048/manual_time_median       2070 ns        10078 ns           20
operands_and_partial_triple_vec<var, double, double>/2048/manual_time_stddev       62.1 ns         86.7 ns           20
operands_and_partial_triple_vec<var, double, double>/4096/manual_time_mean         4339 ns        20431 ns           20
operands_and_partial_triple_vec<var, double, double>/4096/manual_time_median       4265 ns        20403 ns           20
operands_and_partial_triple_vec<var, double, double>/4096/manual_time_stddev        210 ns          270 ns           20
operands_and_partial_triple_vec<var, double, double>/8192/manual_time_mean         8962 ns        41103 ns           20
operands_and_partial_triple_vec<var, double, double>/8192/manual_time_median       8904 ns        40997 ns           20
operands_and_partial_triple_vec<var, double, double>/8192/manual_time_stddev        460 ns          515 ns           20
operands_and_partial_triple_vec<var, var, double>/1024/manual_time_mean            3150 ns        10755 ns           20
operands_and_partial_triple_vec<var, var, double>/1024/manual_time_median          3115 ns        10729 ns           20
operands_and_partial_triple_vec<var, var, double>/1024/manual_time_stddev           260 ns          251 ns           20
operands_and_partial_triple_vec<var, var, double>/2048/manual_time_mean            4445 ns        19390 ns           20
operands_and_partial_triple_vec<var, var, double>/2048/manual_time_median          4425 ns        19368 ns           20
operands_and_partial_triple_vec<var, var, double>/2048/manual_time_stddev           152 ns          177 ns           20
operands_and_partial_triple_vec<var, var, double>/4096/manual_time_mean            9139 ns        40102 ns           20
operands_and_partial_triple_vec<var, var, double>/4096/manual_time_median          9016 ns        39990 ns           20
operands_and_partial_triple_vec<var, var, double>/4096/manual_time_stddev           291 ns          289 ns           20
operands_and_partial_triple_vec<var, var, double>/8192/manual_time_mean           18122 ns        79552 ns           20
operands_and_partial_triple_vec<var, var, double>/8192/manual_time_median         17925 ns        79493 ns           20
operands_and_partial_triple_vec<var, var, double>/8192/manual_time_stddev           654 ns          856 ns           20
operands_and_partial_triple_vec<var, var, var>/1024/manual_time_mean               3413 ns        14443 ns           20
operands_and_partial_triple_vec<var, var, var>/1024/manual_time_median             3384 ns        14436 ns           20
operands_and_partial_triple_vec<var, var, var>/1024/manual_time_stddev             77.6 ns         71.9 ns           20
operands_and_partial_triple_vec<var, var, var>/2048/manual_time_mean               6962 ns        28884 ns           20
operands_and_partial_triple_vec<var, var, var>/2048/manual_time_median             6941 ns        28864 ns           20
operands_and_partial_triple_vec<var, var, var>/2048/manual_time_stddev              241 ns          160 ns           20
operands_and_partial_triple_vec<var, var, var>/4096/manual_time_mean              13954 ns        57573 ns           20
operands_and_partial_triple_vec<var, var, var>/4096/manual_time_median            13871 ns        57547 ns           20
operands_and_partial_triple_vec<var, var, var>/4096/manual_time_stddev              346 ns          343 ns           20
operands_and_partial_triple_vec<var, var, var>/8192/manual_time_mean              27875 ns       113954 ns           20
operands_and_partial_triple_vec<var, var, var>/8192/manual_time_median            27424 ns       113945 ns           20
operands_and_partial_triple_vec<var, var, var>/8192/manual_time_stddev              782 ns          903 ns           20

```


```
-----------------------------------------------------------------------------------------------------------------------
Develop Benchmark                                                                     Time             CPU   Iterations
-----------------------------------------------------------------------------------------------------------------------
operands_and_partial_triple_vec<var, double, double>/1024/manual_time_mean         1336 ns         5466 ns           20
operands_and_partial_triple_vec<var, double, double>/1024/manual_time_median       1332 ns         5471 ns           20
operands_and_partial_triple_vec<var, double, double>/1024/manual_time_stddev       16.9 ns         37.4 ns           20
operands_and_partial_triple_vec<var, double, double>/2048/manual_time_mean         2577 ns        10672 ns           20
operands_and_partial_triple_vec<var, double, double>/2048/manual_time_median       2566 ns        10647 ns           20
operands_and_partial_triple_vec<var, double, double>/2048/manual_time_stddev       59.2 ns         81.4 ns           20
operands_and_partial_triple_vec<var, double, double>/4096/manual_time_mean         5241 ns        21380 ns           20
operands_and_partial_triple_vec<var, double, double>/4096/manual_time_median       5200 ns        21367 ns           20
operands_and_partial_triple_vec<var, double, double>/4096/manual_time_stddev        149 ns          122 ns           20
operands_and_partial_triple_vec<var, double, double>/8192/manual_time_mean        22461 ns        66970 ns           20
operands_and_partial_triple_vec<var, double, double>/8192/manual_time_median      22245 ns        68012 ns           20
operands_and_partial_triple_vec<var, double, double>/8192/manual_time_stddev        456 ns         2494 ns           20
operands_and_partial_triple_vec<var, var, double>/1024/manual_time_mean            2582 ns        10123 ns           20
operands_and_partial_triple_vec<var, var, double>/1024/manual_time_median          2563 ns        10139 ns           20
operands_and_partial_triple_vec<var, var, double>/1024/manual_time_stddev          44.1 ns         72.8 ns           20
operands_and_partial_triple_vec<var, var, double>/2048/manual_time_mean            5199 ns        20182 ns           20
operands_and_partial_triple_vec<var, var, double>/2048/manual_time_median          5160 ns        20200 ns           20
operands_and_partial_triple_vec<var, var, double>/2048/manual_time_stddev          98.3 ns         90.0 ns           20
operands_and_partial_triple_vec<var, var, double>/4096/manual_time_mean           22591 ns        52394 ns           20
operands_and_partial_triple_vec<var, var, double>/4096/manual_time_median         22394 ns        52318 ns           20
operands_and_partial_triple_vec<var, var, double>/4096/manual_time_stddev           425 ns          412 ns           20
operands_and_partial_triple_vec<var, var, double>/8192/manual_time_mean           21757 ns        93674 ns           20
operands_and_partial_triple_vec<var, var, double>/8192/manual_time_median         21653 ns        94365 ns           20
operands_and_partial_triple_vec<var, var, double>/8192/manual_time_stddev           428 ns         2280 ns           20
operands_and_partial_triple_vec<var, var, var>/1024/manual_time_mean               3931 ns        14902 ns           20
operands_and_partial_triple_vec<var, var, var>/1024/manual_time_median             3900 ns        14874 ns           20
operands_and_partial_triple_vec<var, var, var>/1024/manual_time_stddev             68.2 ns          105 ns           20
operands_and_partial_triple_vec<var, var, var>/2048/manual_time_mean              15236 ns        36708 ns           20
operands_and_partial_triple_vec<var, var, var>/2048/manual_time_median            15919 ns        37561 ns           20
operands_and_partial_triple_vec<var, var, var>/2048/manual_time_stddev             1419 ns         1671 ns           20
operands_and_partial_triple_vec<var, var, var>/4096/manual_time_mean              31869 ns        75067 ns           20
operands_and_partial_triple_vec<var, var, var>/4096/manual_time_median            32243 ns        75479 ns           20
operands_and_partial_triple_vec<var, var, var>/4096/manual_time_stddev             1036 ns         1374 ns           20
operands_and_partial_triple_vec<var, var, var>/8192/manual_time_mean              35275 ns       123009 ns           20
operands_and_partial_triple_vec<var, var, var>/8192/manual_time_median            35136 ns       122165 ns           20
operands_and_partial_triple_vec<var, var, var>/8192/manual_time_stddev              637 ns         1763 ns           20
```

So seems good!

Reply:
Eigen vector bench looks good. Scalars look like they might be noisy. Can you also run scalars multiple times like you did vectors? Also please add a bench for `std::vector`.

Reply:
Below are scalars for 20 runs each. PR looks faster on average though has a bit of a higher sd

```
-----------------------------------------------------------------------------------------------------------------
PR Benchmark                                                                    Time             CPU   Iterations
-----------------------------------------------------------------------------------------------------------------
single/manual_time_mean                                 22.7 ns         50.1 ns           20
single/manual_time_median                               22.7 ns         50.0 ns           20
single/manual_time_stddev                              0.113 ns        0.236 ns           20
double<double, var>/manual_time_mean                    22.8 ns         50.4 ns           20
double<double, var>/manual_time_median                  22.8 ns         50.4 ns           20
double<double, var>/manual_time_stddev                 0.074 ns        0.107 ns           20
double<var, double>/manual_time_mean                    22.7 ns         50.2 ns           20
double<var, double>/manual_time_median                  22.7 ns         50.1 ns           20
double<var, double>/manual_time_stddev                 0.083 ns        0.196 ns           20
double<var, var>/manual_time_mean                       24.2 ns         52.1 ns           20
double<var, var>/manual_time_median                     24.1 ns         52.0 ns           20
double<var, var>/manual_time_stddev                    0.167 ns        0.298 ns           20
triple<double, double, double>/manual_time_mean         21.6 ns         49.1 ns           20
triple<double, double, double>/manual_time_median       21.6 ns         49.0 ns           20
triple<double, double, double>/manual_time_stddev      0.156 ns        0.355 ns           20
triple<var, double, double>/manual_time_mean            23.1 ns         51.0 ns           20
triple<var, double, double>/manual_time_median          23.1 ns         51.0 ns           20
triple<var, double, double>/manual_time_stddev         0.064 ns        0.153 ns           20
triple<double, var, double>/manual_time_mean            23.2 ns         51.4 ns           20
triple<double, var, double>/manual_time_median          23.2 ns         51.4 ns           20
triple<double, var, double>/manual_time_stddev         0.075 ns        0.156 ns           20
triple<double, double, var>/manual_time_mean            23.2 ns         51.3 ns           20
triple<double, double, var>/manual_time_median          23.2 ns         51.3 ns           20
triple<double, double, var>/manual_time_stddev         0.048 ns        0.091 ns           20
triple<double, var, var>/manual_time_mean               23.7 ns         52.8 ns           20
triple<double, var, var>/manual_time_median             23.7 ns         52.8 ns           20
triple<double, var, var>/manual_time_stddev            0.144 ns        0.380 ns           20
triple<var, double, var>/manual_time_mean               23.9 ns         52.3 ns           20
triple<var, double, var>/manual_time_median             23.9 ns         52.3 ns           20
triple<var, double, var>/manual_time_stddev            0.028 ns        0.076 ns           20
triple<var, var, double>/manual_time_mean               24.3 ns         52.4 ns           20
triple<var, var, double>/manual_time_median             24.3 ns         52.3 ns           20
triple<var, var, double>/manual_time_stddev            0.143 ns        0.278 ns           20
triple<var, var, var>/manual_time_mean                  26.3 ns         55.3 ns           20
triple<var, var, var>/manual_time_median                26.2 ns         55.3 ns           20
triple<var, var, var>/manual_time_stddev               0.201 ns        0.219 ns           20
```

```
-----------------------------------------------------------------------------------------------------------------
Develop Benchmark                                                               Time             CPU   Iterations
-----------------------------------------------------------------------------------------------------------------
single/manual_time_mean                                 23.6 ns         52.3 ns           20
single/manual_time_median                               23.6 ns         52.3 ns           20
single/manual_time_stddev                              0.096 ns        0.412 ns           20
double<double, var>/manual_time_mean                    24.3 ns         52.7 ns           20
double<double, var>/manual_time_median                  24.2 ns         52.5 ns           20
double<double, var>/manual_time_stddev                 0.433 ns        0.905 ns           20
double<var, double>/manual_time_mean                    24.5 ns         53.0 ns           20
double<var, double>/manual_time_median                  24.5 ns         53.1 ns           20
double<var, double>/manual_time_stddev                 0.254 ns        0.596 ns           20
double<var, var>/manual_time_mean                       25.4 ns         54.9 ns           20
double<var, var>/manual_time_median                     25.4 ns         54.9 ns           20
double<var, var>/manual_time_stddev                    0.095 ns        0.210 ns           20
triple<double, double, double>/manual_time_mean         22.0 ns         50.0 ns           20
triple<double, double, double>/manual_time_median       22.1 ns         50.2 ns           20
triple<double, double, double>/manual_time_stddev      0.242 ns        0.516 ns           20
triple<var, double, double>/manual_time_mean            23.8 ns         52.3 ns           20
triple<var, double, double>/manual_time_median          23.8 ns         52.2 ns           20
triple<var, double, double>/manual_time_stddev         0.125 ns        0.280 ns           20
triple<double, var, double>/manual_time_mean            24.5 ns         53.0 ns           20
triple<double, var, double>/manual_time_median          24.5 ns         53.1 ns           20
triple<double, var, double>/manual_time_stddev         0.200 ns        0.472 ns           20
triple<double, double, var>/manual_time_mean            24.5 ns         53.0 ns           20
triple<double, double, var>/manual_time_median          24.4 ns         52.9 ns           20
triple<double, double, var>/manual_time_stddev         0.057 ns        0.153 ns           20
triple<double, var, var>/manual_time_mean               26.1 ns         55.6 ns           20
triple<double, var, var>/manual_time_median             26.1 ns         55.6 ns           20
triple<double, var, var>/manual_time_stddev            0.231 ns        0.338 ns           20
triple<var, double, var>/manual_time_mean               26.1 ns         55.6 ns           20
triple<var, double, var>/manual_time_median             26.1 ns         55.6 ns           20
triple<var, double, var>/manual_time_stddev            0.155 ns        0.224 ns           20
triple<var, var, double>/manual_time_mean               25.3 ns         54.6 ns           20
triple<var, var, double>/manual_time_median             25.3 ns         54.6 ns           20
triple<var, var, double>/manual_time_stddev            0.090 ns        0.160 ns           20
triple<var, var, var>/manual_time_mean                  27.0 ns         56.9 ns           20
triple<var, var, var>/manual_time_median                27.0 ns         56.9 ns           20
triple<var, var, var>/manual_time_stddev               0.070 ns        0.149 ns           20
```

Reply:
Good catch on the std vector check! So the way I originally had this the PR's std vector stuff was actually slower

```cpp
-----------------------------------------------------------------------------------------------------------------------
Old PR Benchmark                                                Time             CPU   Iterations
-----------------------------------------------------------------------------------------------------------------------
triple_vec<var, double, double>/1024/manual_time_mean         1438 ns         6162 ns           20
triple_vec<var, double, double>/1024/manual_time_median       1440 ns         5838 ns           20
triple_vec<var, double, double>/1024/manual_time_stddev       26.0 ns          696 ns           20
triple_vec<var, double, double>/2048/manual_time_mean         2718 ns        11288 ns           20
triple_vec<var, double, double>/2048/manual_time_median       2711 ns        11283 ns           20
triple_vec<var, double, double>/2048/manual_time_stddev       23.8 ns         55.0 ns           20
triple_vec<var, double, double>/4096/manual_time_mean         5970 ns        23373 ns           20
triple_vec<var, double, double>/4096/manual_time_median       5878 ns        23214 ns           20
triple_vec<var, double, double>/4096/manual_time_stddev        389 ns          610 ns           20
triple_vec<var, double, double>/8192/manual_time_mean        12972 ns        47975 ns           20
triple_vec<var, double, double>/8192/manual_time_median      12908 ns        47749 ns           20
triple_vec<var, double, double>/8192/manual_time_stddev        580 ns         1044 ns           20
triple_vec<var, var, double>/1024/manual_time_mean            3070 ns        11080 ns           20
triple_vec<var, var, double>/1024/manual_time_median          3056 ns        11091 ns           20
triple_vec<var, var, double>/1024/manual_time_stddev          46.9 ns         78.1 ns           20
triple_vec<var, var, double>/2048/manual_time_mean            5921 ns        21778 ns           20
triple_vec<var, var, double>/2048/manual_time_median          5883 ns        21768 ns           20
triple_vec<var, var, double>/2048/manual_time_stddev          98.7 ns          153 ns           20
triple_vec<var, var, double>/4096/manual_time_mean           13163 ns        44912 ns           20
triple_vec<var, var, double>/4096/manual_time_median         13137 ns        44861 ns           20
triple_vec<var, var, double>/4096/manual_time_stddev           206 ns          356 ns           20
triple_vec<var, var, double>/8192/manual_time_mean           23865 ns       102295 ns           20
triple_vec<var, var, double>/8192/manual_time_median         23680 ns       102238 ns           20
triple_vec<var, var, double>/8192/manual_time_stddev           302 ns          526 ns           20
triple_vec<var, var, var>/1024/manual_time_mean               4064 ns        15433 ns           20
triple_vec<var, var, var>/1024/manual_time_median             4045 ns        15386 ns           20
triple_vec<var, var, var>/1024/manual_time_stddev             70.8 ns          150 ns           20
triple_vec<var, var, var>/2048/manual_time_mean               8072 ns        30585 ns           20
triple_vec<var, var, var>/2048/manual_time_median             8021 ns        30562 ns           20
triple_vec<var, var, var>/2048/manual_time_stddev              116 ns          148 ns           20
triple_vec<var, var, var>/4096/manual_time_mean              17212 ns        62197 ns           20
triple_vec<var, var, var>/4096/manual_time_median            17172 ns        62281 ns           20
triple_vec<var, var, var>/4096/manual_time_stddev              118 ns          306 ns           20
triple_vec<var, var, var>/8192/manual_time_mean              35604 ns       125238 ns           20
triple_vec<var, var, var>/8192/manual_time_median            35575 ns       125105 ns           20
triple_vec<var, var, var>/8192/manual_time_stddev              359 ns          559 ns           20
```

```cpp
-----------------------------------------------------------------------------------------------------------------------
Develop Benchmark                                                Time             CPU   Iterations
-----------------------------------------------------------------------------------------------------------------------
triple_vec<var, double, double>/1024/manual_time_mean         1274 ns         5570 ns           30
triple_vec<var, double, double>/1024/manual_time_median       1270 ns         5580 ns           30
triple_vec<var, double, double>/1024/manual_time_stddev       16.0 ns         34.0 ns           30
triple_vec<var, double, double>/2048/manual_time_mean         2575 ns        11261 ns           30
triple_vec<var, double, double>/2048/manual_time_median       2581 ns        11296 ns           30
triple_vec<var, double, double>/2048/manual_time_stddev       65.4 ns          204 ns           30
triple_vec<var, double, double>/4096/manual_time_mean         5372 ns        22849 ns           30
triple_vec<var, double, double>/4096/manual_time_median       5301 ns        22713 ns           30
triple_vec<var, double, double>/4096/manual_time_stddev        251 ns          480 ns           30
triple_vec<var, double, double>/8192/manual_time_mean        23579 ns        58686 ns           30
triple_vec<var, double, double>/8192/manual_time_median      23413 ns        58610 ns           30
triple_vec<var, double, double>/8192/manual_time_stddev        308 ns          489 ns           30
triple_vec<var, var, double>/1024/manual_time_mean            2747 ns        10699 ns           30
triple_vec<var, var, double>/1024/manual_time_median          2725 ns        10701 ns           30
triple_vec<var, var, double>/1024/manual_time_stddev          53.8 ns         79.3 ns           30
triple_vec<var, var, double>/2048/manual_time_mean            5626 ns        21473 ns           30
triple_vec<var, var, double>/2048/manual_time_median          5601 ns        21490 ns           30
triple_vec<var, var, double>/2048/manual_time_stddev          89.3 ns          145 ns           30
triple_vec<var, var, double>/4096/manual_time_mean           23963 ns        55617 ns           30
triple_vec<var, var, double>/4096/manual_time_median         23897 ns        55686 ns           30
triple_vec<var, var, double>/4096/manual_time_stddev           500 ns          669 ns           30
triple_vec<var, var, double>/8192/manual_time_mean           23946 ns        87436 ns           30
triple_vec<var, var, double>/8192/manual_time_median         23790 ns        87511 ns           30
triple_vec<var, var, double>/8192/manual_time_stddev           368 ns          560 ns           30
triple_vec<var, var, var>/1024/manual_time_mean               4137 ns        15595 ns           30
triple_vec<var, var, var>/1024/manual_time_median             4109 ns        15596 ns           30
triple_vec<var, var, var>/1024/manual_time_stddev             56.7 ns         92.0 ns           30
triple_vec<var, var, var>/2048/manual_time_mean               9011 ns        31830 ns           30
triple_vec<var, var, var>/2048/manual_time_median             8821 ns        31642 ns           30
triple_vec<var, var, var>/2048/manual_time_stddev              661 ns          614 ns           30
triple_vec<var, var, var>/4096/manual_time_mean              17588 ns        62469 ns           30
triple_vec<var, var, var>/4096/manual_time_median            17437 ns        62487 ns           30
triple_vec<var, var, var>/4096/manual_time_stddev              424 ns          480 ns           30
triple_vec<var, var, var>/8192/manual_time_mean              35678 ns       125320 ns           30
triple_vec<var, var, var>/8192/manual_time_median            35429 ns       125398 ns           30
triple_vec<var, var, var>/8192/manual_time_stddev              613 ns          868 ns           30
```

Looking into it more, I think it's because of the `operand()` and `partial()` functions that return vectors returning by reference, which when we copy into the lambda is going to call the copy constructor for each vec. Since `std::vector` doesn't know it can just do the equivalent of a move and give the new vector it's already allocated memory. So I changed the `operand()` and `partial()` getters to perform a move like 

```cpp
inline auto&& operand() noexcept { return std::move(this->operands_); }
```

and that seems to make the new impl faster

```cpp
-----------------------------------------------------------------------------------------------------------------------
New PR Benchmark                                               Time             CPU   Iterations
-----------------------------------------------------------------------------------------------------------------------
triple_vec<var, double, double>/128/manual_time_mean           223 ns          830 ns           30
triple_vec<var, double, double>/128/manual_time_median         223 ns          832 ns           30
triple_vec<var, double, double>/128/manual_time_stddev        2.59 ns         8.00 ns           30
triple_vec<var, double, double>/256/manual_time_mean           379 ns         1558 ns           30
triple_vec<var, double, double>/256/manual_time_median         376 ns         1559 ns           30
triple_vec<var, double, double>/256/manual_time_stddev        6.13 ns         14.0 ns           30
triple_vec<var, double, double>/512/manual_time_mean           706 ns         2967 ns           30
triple_vec<var, double, double>/512/manual_time_median         703 ns         2968 ns           30
triple_vec<var, double, double>/512/manual_time_stddev        10.5 ns         17.1 ns           30
triple_vec<var, double, double>/1024/manual_time_mean         1365 ns         6162 ns           30
triple_vec<var, double, double>/1024/manual_time_median       1353 ns         5854 ns           30
triple_vec<var, double, double>/1024/manual_time_stddev       27.9 ns          709 ns           30
triple_vec<var, double, double>/2048/manual_time_mean         2687 ns        13986 ns           30
triple_vec<var, double, double>/2048/manual_time_median       2674 ns        15219 ns           30
triple_vec<var, double, double>/2048/manual_time_stddev       36.4 ns         1817 ns           30
triple_vec<var, double, double>/4096/manual_time_mean         5686 ns        23427 ns           30
triple_vec<var, double, double>/4096/manual_time_median       5507 ns        23216 ns           30
triple_vec<var, double, double>/4096/manual_time_stddev        419 ns          608 ns           30
triple_vec<var, double, double>/8192/manual_time_mean        11470 ns        61792 ns           30
triple_vec<var, double, double>/8192/manual_time_median      11401 ns        61799 ns           30
triple_vec<var, double, double>/8192/manual_time_stddev        530 ns          435 ns           30
triple_vec<var, var, double>/128/manual_time_mean              387 ns         1416 ns           30
triple_vec<var, var, double>/128/manual_time_median            386 ns         1420 ns           30
triple_vec<var, var, double>/128/manual_time_stddev           8.53 ns         12.2 ns           30
triple_vec<var, var, double>/256/manual_time_mean              655 ns         2744 ns           30
triple_vec<var, var, double>/256/manual_time_median            651 ns         2746 ns           30
triple_vec<var, var, double>/256/manual_time_stddev           17.0 ns         18.1 ns           30
triple_vec<var, var, double>/512/manual_time_mean             1194 ns         5234 ns           30
triple_vec<var, var, double>/512/manual_time_median           1190 ns         5244 ns           30
triple_vec<var, var, double>/512/manual_time_stddev           24.2 ns         31.3 ns           30
triple_vec<var, var, double>/1024/manual_time_mean            2348 ns        10330 ns           30
triple_vec<var, var, double>/1024/manual_time_median          2326 ns        10341 ns           30
triple_vec<var, var, double>/1024/manual_time_stddev          46.9 ns         73.0 ns           30
triple_vec<var, var, double>/2048/manual_time_mean            4548 ns        20499 ns           30
triple_vec<var, var, double>/2048/manual_time_median          4526 ns        20533 ns           30
triple_vec<var, var, double>/2048/manual_time_stddev          79.2 ns          138 ns           30
triple_vec<var, var, double>/4096/manual_time_mean           16114 ns        55232 ns           30
triple_vec<var, var, double>/4096/manual_time_median         16068 ns        55381 ns           30
triple_vec<var, var, double>/4096/manual_time_stddev           391 ns          657 ns           30
triple_vec<var, var, double>/8192/manual_time_mean           19139 ns        97947 ns           30
triple_vec<var, var, double>/8192/manual_time_median         18875 ns        98013 ns           30
triple_vec<var, var, double>/8192/manual_time_stddev           611 ns          642 ns           30
triple_vec<var, var, var>/128/manual_time_mean                 590 ns         2702 ns           30
triple_vec<var, var, var>/128/manual_time_median               591 ns         2781 ns           30
triple_vec<var, var, var>/128/manual_time_stddev              13.2 ns          330 ns           30
triple_vec<var, var, var>/256/manual_time_mean                1055 ns         5403 ns           30
triple_vec<var, var, var>/256/manual_time_median              1051 ns         5517 ns           30
triple_vec<var, var, var>/256/manual_time_stddev              15.8 ns          672 ns           30
triple_vec<var, var, var>/512/manual_time_mean                1947 ns        10263 ns           30
triple_vec<var, var, var>/512/manual_time_median              1940 ns        10201 ns           30
triple_vec<var, var, var>/512/manual_time_stddev              18.0 ns         1339 ns           30
triple_vec<var, var, var>/1024/manual_time_mean               3721 ns        18748 ns           30
triple_vec<var, var, var>/1024/manual_time_median             3700 ns        18016 ns           30
triple_vec<var, var, var>/1024/manual_time_stddev             40.9 ns         2908 ns           30
triple_vec<var, var, var>/2048/manual_time_mean               7337 ns        43676 ns           30
triple_vec<var, var, var>/2048/manual_time_median             7309 ns        44022 ns           30
triple_vec<var, var, var>/2048/manual_time_stddev             64.7 ns         5479 ns           30
triple_vec<var, var, var>/4096/manual_time_mean              19954 ns        96306 ns           30
triple_vec<var, var, var>/4096/manual_time_median            20378 ns       103849 ns           30
triple_vec<var, var, var>/4096/manual_time_stddev             1299 ns        12990 ns           30
triple_vec<var, var, var>/8192/manual_time_mean              31356 ns       159140 ns           30
triple_vec<var, var, var>/8192/manual_time_median            31218 ns       159218 ns           30
triple_vec<var, var, var>/8192/manual_time_stddev              457 ns        22305 ns           30
```

```cpp
-----------------------------------------------------------------------------------------------------------------------
Develop Benchmark                                                Time             CPU   Iterations
-----------------------------------------------------------------------------------------------------------------------
triple_vec<var, double, double>/128/manual_time_mean           241 ns          854 ns           30
triple_vec<var, double, double>/128/manual_time_median         244 ns          857 ns           30
triple_vec<var, double, double>/128/manual_time_stddev        8.66 ns         11.9 ns           30
triple_vec<var, double, double>/256/manual_time_mean           371 ns         1551 ns           30
triple_vec<var, double, double>/256/manual_time_median         370 ns         1550 ns           30
triple_vec<var, double, double>/256/manual_time_stddev        6.11 ns         14.9 ns           30
triple_vec<var, double, double>/512/manual_time_mean           684 ns         2916 ns           30
triple_vec<var, double, double>/512/manual_time_median         681 ns         2916 ns           30
triple_vec<var, double, double>/512/manual_time_stddev        8.67 ns         19.3 ns           30
triple_vec<var, double, double>/1024/manual_time_mean         1274 ns         5570 ns           30
triple_vec<var, double, double>/1024/manual_time_median       1270 ns         5580 ns           30
triple_vec<var, double, double>/1024/manual_time_stddev       16.0 ns         34.0 ns           30
triple_vec<var, double, double>/2048/manual_time_mean         2575 ns        11261 ns           30
triple_vec<var, double, double>/2048/manual_time_median       2581 ns        11296 ns           30
triple_vec<var, double, double>/2048/manual_time_stddev       65.4 ns          204 ns           30
triple_vec<var, double, double>/4096/manual_time_mean         5372 ns        22849 ns           30
triple_vec<var, double, double>/4096/manual_time_median       5301 ns        22713 ns           30
triple_vec<var, double, double>/4096/manual_time_stddev        251 ns          480 ns           30
triple_vec<var, double, double>/8192/manual_time_mean        23579 ns        58686 ns           30
triple_vec<var, double, double>/8192/manual_time_median      23413 ns        58610 ns           30
triple_vec<var, double, double>/8192/manual_time_stddev        308 ns          489 ns           30
triple_vec<var, var, double>/128/manual_time_mean              480 ns         1563 ns           30
triple_vec<var, var, double>/128/manual_time_median            491 ns         1574 ns           30
triple_vec<var, var, double>/128/manual_time_stddev           25.0 ns         32.6 ns           30
triple_vec<var, var, double>/256/manual_time_mean              775 ns         2870 ns           30
triple_vec<var, var, double>/256/manual_time_median            772 ns         2869 ns           30
triple_vec<var, var, double>/256/manual_time_stddev           17.1 ns         31.7 ns           30
triple_vec<var, var, double>/512/manual_time_mean             1427 ns         5467 ns           30
triple_vec<var, var, double>/512/manual_time_median           1418 ns         5470 ns           30
triple_vec<var, var, double>/512/manual_time_stddev           19.4 ns         32.9 ns           30
triple_vec<var, var, double>/1024/manual_time_mean            2747 ns        10699 ns           30
triple_vec<var, var, double>/1024/manual_time_median          2725 ns        10701 ns           30
triple_vec<var, var, double>/1024/manual_time_stddev          53.8 ns         79.3 ns           30
triple_vec<var, var, double>/2048/manual_time_mean            5626 ns        21473 ns           30
triple_vec<var, var, double>/2048/manual_time_median          5601 ns        21490 ns           30
triple_vec<var, var, double>/2048/manual_time_stddev          89.3 ns          145 ns           30
triple_vec<var, var, double>/4096/manual_time_mean           23963 ns        55617 ns           30
triple_vec<var, var, double>/4096/manual_time_median         23897 ns        55686 ns           30
triple_vec<var, var, double>/4096/manual_time_stddev           500 ns          669 ns           30
triple_vec<var, var, double>/8192/manual_time_mean           23946 ns        87436 ns           30
triple_vec<var, var, double>/8192/manual_time_median         23790 ns        87511 ns           30
triple_vec<var, var, double>/8192/manual_time_stddev           368 ns          560 ns           30
triple_vec<var, var, var>/128/manual_time_mean                 702 ns         2241 ns           30
triple_vec<var, var, var>/128/manual_time_median               711 ns         2244 ns           30
triple_vec<var, var, var>/128/manual_time_stddev              25.6 ns         27.7 ns           30
triple_vec<var, var, var>/256/manual_time_mean                1159 ns         4154 ns           30
triple_vec<var, var, var>/256/manual_time_median              1154 ns         4166 ns           30
triple_vec<var, var, var>/256/manual_time_stddev              12.6 ns         28.0 ns           30
triple_vec<var, var, var>/512/manual_time_mean                2117 ns         7923 ns           30
triple_vec<var, var, var>/512/manual_time_median              2107 ns         7923 ns           30
triple_vec<var, var, var>/512/manual_time_stddev              26.4 ns         55.8 ns           30
triple_vec<var, var, var>/1024/manual_time_mean               4137 ns        15595 ns           30
triple_vec<var, var, var>/1024/manual_time_median             4109 ns        15596 ns           30
triple_vec<var, var, var>/1024/manual_time_stddev             56.7 ns         92.0 ns           30
triple_vec<var, var, var>/2048/manual_time_mean               9011 ns        31830 ns           30
triple_vec<var, var, var>/2048/manual_time_median             8821 ns        31642 ns           30
triple_vec<var, var, var>/2048/manual_time_stddev              661 ns          614 ns           30
triple_vec<var, var, var>/4096/manual_time_mean              17588 ns        62469 ns           30
triple_vec<var, var, var>/4096/manual_time_median            17437 ns        62487 ns           30
triple_vec<var, var, var>/4096/manual_time_stddev              424 ns          480 ns           30
triple_vec<var, var, var>/8192/manual_time_mean              35678 ns       125320 ns           30
triple_vec<var, var, var>/8192/manual_time_median            35429 ns       125398 ns           30
triple_vec<var, var, var>/8192/manual_time_stddev              613 ns          868 ns           30
```

It seems like there is still some overhead at small N but for larger N it looks like avoiding that copy is enough to give a little speed bump

Reply:
Can you check if you can do anything about performance of small vectors? Current state looks good enough that I can approve it, but it would be really nice if we would not slow down any use case.

Reply:
I'm not sure if I can, and eyeballing the chart it looks like the smaller std vector versions that are a bit slower are still within 1-2 std of the original